### PR TITLE
Add AutoCache Primitive

### PR DESCRIPTION
**Pull request by Bastani**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -266,6 +266,39 @@ autoMap.Remove("Pickles");
 autoMap["Brisket"] = new Cat("Brisket");
 ```
 
+## 💰 AutoCache
+
+`AutoCache` is a cache which stores values separated by type. On update, it broadcasts to all subscribers and stores the value based on the type given.
```
Comment (jolexxa) [inline README.md]: Since this is less common than the other reactive primitives, we'll need to outline potential use cases for when you might use something like this.

File: README.md [hunk 2]
```diff
@@ -266,6 +266,39 @@ autoMap.Remove("Pickles");
...
+## 💰 AutoCache
+
+`AutoCache` is a cache which stores values separated by type. On update, it broadcasts to all subscribers and stores the value based on the type given. 
+You can then use the method `TryGetValue<T>(out T value)` to get the last value updated of type `T`.
+
+> [!NOTE]
+> While AutoCache does support reference types, consider using value types instead when initializing new instances on update. 
+
+```csharp
+readonly record struct UpdateName(string DogName);
```
Comment (jolexxa) [inline README.md:286]: This code example doesn't demonstrate sending multiple different value types and multiple reference types that extend from the same type — we want to showcase the real power and flexibility this offers.

File: README.md [hunk 3]
```diff
@@ -266,6 +266,39 @@ autoMap.Remove("Pickles");
 autoMap["Brisket"] = new Cat("Brisket");
 ```
 
+## 💰 AutoCache
+
+`AutoCache` is a cache which stores values separated by type. On update, it broadcasts to all subscribers and stores the value based on the type given. 
+You can then use the method `TryGetValue<T>(out T value)` to get the last value updated of type `T`.
```
Comment (jolexxa) [inline README.md]: We also want a `[!CAUTION]` block aside that that demonstrates you must call `TryGetValue` with the type that was pushed and demonstrate how it would fail if you tried accessing it by a different type.

File: README.md (hunk 1 above)
Comment (jolexxa) [inline README.md]: We want to mention that because this is not generic, it can be used as a message channel and lookup-cache for multiple types of data. We also want to make it very clear that it is optimized for value types and will not box them with every update.

File: Chickensoft.Sync/src/primitives/AutoCache.cs [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+    HasValue = false;
+  }
+}
+
+/// <summary>
+/// <para>
+/// A cache which stores values separated by type.
+/// </para>
+/// <para>
+/// On update, it broadcasts to all subscribers and stores the value based on
```
Comment (jolexxa) [inline Chickensoft.Sync/src/primitives/AutoCache.cs]: Subscribers is fine, but it may be better to use the term `binding` for this library.

File: Chickensoft.Sync/src/primitives/AutoCache.cs [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+    /// <returns>This binding (for chaining).</returns>
+    /// <seealso cref="OnUpdate{T}(Callback{T}, Func{T, bool}?)"/>
+    public Binding OnUpdate<T>(
+      Action<T> callback, Func<T, bool>? condition = null) where T : class
+    {
+      bool predicate(T value) => condition?.Invoke(value) ?? true;
+
+      AddCallback(
+        (in RefValue value) => {
+          if (value.Value is T tValue) {
```
Comment (jolexxa) [inline Chickensoft.Sync/src/primitives/AutoCache.cs]: Since you're passing in a predicate on 116 that verifies the type, we can probably use `Unsafe.As` to cast without penalty here instead of branching again.

File: Chickensoft.Sync/src/primitives/AutoCache.cs [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    _valueDict.Clear();
+  }
+
+  /// <inheritdoc />
+  /// <returns>true if the <see cref="AutoCache"/> contains an element with the
+  /// specified type; otherwise, false.</returns>
+  public bool TryGetValue<T>([MaybeNullWhen(false)] out T value) {
+    value = default;
+    if (_valueDict.TryGetValue(typeof(T), out var val) &&
+        val is CachedValue<T> { HasValue: true } derivedValue) {
```
Comment (jolexxa) [inline Chickensoft.Sync/src/primitives/AutoCache.cs:177]: we love a good pattern matcher

File: Chickensoft.Sync/src/primitives/AutoCache.cs [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+      value = derivedValue.Value!;
+	    return true;
+    }
+    if (_refDict.TryGetValue(typeof(T), out var refVal) &&
+        refVal is T derivedRef) {
```
Comment (jolexxa) [inline Chickensoft.Sync/src/primitives/AutoCache.cs:182]: I believe `refVal is T` will fail on `null` for reference types so we may want to add the `notnull` constraint to the type parameters where relevant — not sure how that will play out off the top of my head but worth a shot. Also probably want to test for `null` handling for reference types in a unit test.

**Review by jolexxa — CHANGES_REQUESTED** [review]
This is absolutely beautiful and shaping up to be a work of art 💯 Just a bunch of small things and documentation requests. Thank you soooo much. This is an incredibly powerful construct and will hopefully benefit a lot of folks in the long term.

**Comment by jolexxa** [comment]
@Bastani You'll need to add the spelling exceptions to the cspell.json ;)

File: Chickensoft.Sync/src/primitives/AutoCache.cs [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+    /// <returns>This binding (for chaining).</returns>
+    /// <seealso cref="OnUpdate{T}(Callback{T}, Func{T, bool}?)"/>
+    public Binding OnUpdate<T>(
+      Action<T> callback, Func<T, bool>? condition = null) where T : class
+    {
+      bool predicate(T value) => condition?.Invoke(value) ?? true;
+
+      AddCallback(
+        (in RefValue value) => {
+            callback(Unsafe.As<T>(value.Value));
```
Comment (jolexxa) [inline Chickensoft.Sync/src/primitives/AutoCache.cs]: You might be able to just expression body this

File: README.md [hunk 9]
```diff
@@ -266,6 +266,59 @@ autoMap.Remove("Pickles");
...
+  Console.WriteLine($"Last received dog name: {update.DogName}"
+}
+
+binding
+  .OnUpdate<Animal>(
+    (animal) => Console.WriteLine($"Animal Updated: {animal.Name}")
+  );
+
+autoCache.Update<Animal>(new Mouse("Hamtaro"));
+autoCache.Update(new Dog("Chibi"));
```
Comment (jolexxa) [inline README.md]:
```suggestion
autoCache.Update(new Dog("Cookie"));
```

Silly, but Chibi is one of my friends' cats — Cookie is one of my dogs.

File: README.md [hunk 10]
```diff
@@ -266,6 +266,59 @@ autoMap.Remove("Pickles");
...
+  // This would print out "Last received dog name: Sven"
+  Console.WriteLine($"Last received dog name: {update.DogName}"
+}
+
+binding
+  .OnUpdate<Animal>(
+    (animal) => Console.WriteLine($"Animal Updated: {animal.Name}")
+  );
+
+autoCache.Update<Animal>(new Mouse("Hamtaro"));
```
Comment (jolexxa) [inline README.md:312]:
```suggestion
// Store and broadcast a Mouse by its less-specific supertype, Animal
autoCache.Update<Animal>(new Mouse("Hamtaro"));
```

File: README.md [hunk 11]
```diff
@@ -266,6 +266,59 @@ autoMap.Remove("Pickles");
...
+for multiple types of data. We've optimized AutoCache for value types so that it does not box value types on updates. 
+You might find this pattern familiar if you've used Chickensoft.LogicBlocks.
+
+> [!CAUTION]
+> When pushing a value of type `Dog` which derives from Animal, TryGetValue<Animal> will not return the last value updated of type `Dog`.
+> If you desire to get the last Animal value updated, you will have to use Update<Animal>(new Dog()) instead.
+> Although Binding notifications for OnUpdate<Dog> or OnUpdate<Animal> will still be called regardless of the type pushed.
+
+> [!NOTE]
+> While AutoCache does support reference types, consider using value types instead when initializing new instances on update.
```
Comment (jolexxa) [inline README.md]:
```suggestion
> While AutoCache does support reference types, consider using value types instead when initializing new instances on update to avoid allocating unnecessary memory that would need to be immediately collected by the garbage collector. Using value types where possible helps avoid stuttering and hitches by reducing the amount of work that the garbage collector needs to do to clean up reference types on the heap.
```

File: Chickensoft.Sync/src/primitives/AutoCache.cs [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+  /// <summary>
+  /// The combined count of all reference types and value types stored in the cache.
+  /// </summary>
+  public int Count => _refDict.Count + _cacheCount;
+
+  /// <summary>
+  /// Clears the cache of any stored values.
+  /// </summary>
+  public void Clear() {
+    _refDict.Clear();
```
Comment (jolexxa) [inline Chickensoft.Sync/src/primitives/AutoCache.cs]: We'll need to introduce a new atomic operation for Clear and implement the actual clearing in `IPerform<ClearOp>.Perform(in ClearOp op)` — otherwise, this will not behave deterministically. A unit test to guarantee clear is deferred (clearing from a binding callback) would suffice to prevent a regression here, too.

File: Chickensoft.Sync/src/primitives/AutoCache.cs [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using Collections;
+using Sync;
+
+internal abstract class CachedValue {
+  public abstract void Clear();
+}
+
+internal class CachedValue<T> : CachedValue {
```
Comment (jolexxa) [inline Chickensoft.Sync/src/primitives/AutoCache.cs:14]: It's funny because we've created https://learn.microsoft.com/en-us/dotnet/api/system.nullable-1 without the `struct` constraint on `T`.

**Review by jolexxa — CHANGES_REQUESTED** [review]
Awesome, thank you. Looking really really good. I spotted a determinism problem I didn't catch before and added a few suggestions for docs and comments.

**Review by jolexxa — APPROVED** [review]
